src/lazy/expanded/mod.rs

```diff
@@ -131,84 +134,334 @@ impl<'top, 'data, D: LazyDecoder<'data>> ExpandedStreamItem<'top, 'data, D> {
+                    };
                     // Push the invocation onto the evaluation stack.
-                    self.evaluator.push(context, e_exp)?;
-                    // Return to the top of the loop to pull the next value (if any) from the evaluator.
-                    continue;
+                    evaluator.push(context, resolved_e_exp)?;
+                    self.evaluator_ptr
+                        .set(Some(Self::evaluator_to_ptr(evaluator)));
+
+                    // Try to get a value by starting to evaluate the e-expression.
+                    if let Some(value) = self.next_from_evaluator()? {
+                        // If we get a value, return it.
+                        return Ok(value);
+                    } else {
+                        // If the expression was equivalent to `(:void)`, return to the top of
+                        // the loop and get the next expression.
+                        continue;
+                    }
                 }
-                EndOfStream => ExpandedStreamItem::EndOfStream,
+                EndOfStream => return Ok(SystemStreamItem::EndOfStream),
             };
-            return Ok(expanded_item);
+        }
+    }
+
+    /// If there is not an evaluation in process, returns `Ok(None)`.
+    /// If there is an evaluation in process but it does not yield another value, returns `Ok(None)`.
+    /// If there is an evaluation in process and it yields another value, returns `Ok(Some(value))`.
+    /// Otherwise, returns `Err`.
+    fn next_from_evaluator<'top>(
+        &'top self,
+    ) -> IonResult<Option<SystemStreamItem<'top, 'data, D>>> {
+        let evaluator_ptr = match self.evaluator_ptr.get() {
+            // There's not currently an evaluator.
+            None => return Ok(None),
+            // There's an evaluator in the process of expanding a macro.
+            Some(ptr) => ptr,
+        };
+        let evaluator = Self::ptr_to_evaluator(evaluator_ptr);
+
+        match evaluator.next(self.context(), 0) {
+            Ok(Some(value)) => {
+                // See if this value was a symbol table that needs interpretation.
+                self.interpret_value(value).map(Some)
+            }
+            Ok(None) => {
+                // While the evaluator had macros in its stack, they did not produce any more
+                // values. The stack is now empty.
+                Ok(None)
+            }
+            Err(e) => Err(e),
         }
     }
 }
 
 /// The source of data backing a [`LazyExpandedValue`].
-#[derive(Debug, Clone)]
+#[derive(Copy, Clone)]
 pub enum ExpandedValueSource<'top, 'data, D: LazyDecoder<'data>> {
     /// This value was a literal in the input stream.
     ValueLiteral(D::Value),
     /// This value was part of a template definition.
-    Template(&'top Element),
-    /// This value was the computed result of a macro invocation like `(:make_string ...)`.
+    Template(Environment<'top, 'data, D>, TemplateElement<'top>),
```

Comment:
🗺️ We no longer use an `Element` (which allocates on the heap and uses atomic operations) as our template value. Instead, we use `TemplateElement`, which is either a plain scalar value or a container type holding a number of child elements.